packages/next/lib/typescript/writeAppTypeDeclarations.ts

```diff
@@ -9,19 +10,27 @@ export async function writeAppTypeDeclarations(
   // Reference `next` types
   const appTypeDeclarations = path.join(baseDir, 'next-env.d.ts')
 
-  await fs.writeFile(
-    appTypeDeclarations,
+  const content =
     '/// <reference types="next" />' +
-      os.EOL +
-      '/// <reference types="next/types/global" />' +
-      os.EOL +
-      (imageImportsEnabled
-        ? '/// <reference types="next/image-types/global" />' + os.EOL
-        : '') +
-      os.EOL +
-      '// NOTE: This file should not be edited' +
-      os.EOL +
-      '// see https://nextjs.org/docs/basic-features/typescript for more information.' +
-      os.EOL
-  )
+    os.EOL +
+    '/// <reference types="next/types/global" />' +
+    os.EOL +
+    (imageImportsEnabled
+      ? '/// <reference types="next/image-types/global" />' + os.EOL
+      : '') +
+    os.EOL +
+    '// NOTE: This file should not be edited' +
+    os.EOL +
+    '// see https://nextjs.org/docs/basic-features/typescript for more information.' +
+    os.EOL
+
+  // Avoids a write for read-only filesystems
+  if (
+    (await fileExists(appTypeDeclarations)) &&
+    (await fs.readFile(appTypeDeclarations, 'utf8')) === content
```

Comment:
```suggestion
    (await fs.readFile(appTypeDeclarations, 'utf8').catch(() => '') === content
```
Instead of checking if the file exists and then reading the content we could read the content and handle it failing gracefully which seems to be recommended in the node docs [here](https://nodejs.org/api/fs.html#fs_fs_exists_path_callback)

Reply:
Ah. That's a good point and would have fixed it if the merge queue hadn't kicked in. I think practically, that case is unlikely to be an issue here - the window where the user would have to write the file is slim and the consequence is small. Is it worth a PR to fix?

Reply:
Yeah makes sense, I think it's alright to leave as is